hebiros_gazebo_plugin/plugin_hebiros/hebiros_gazebo_group.cpp

```diff
@@ -0,0 +1,237 @@
+    this->check_acknowledgement = false;
+  }
+
+  ros::Time current_time = ros::Time::now();
+  this->start_time = current_time;
+  this->prev_time = current_time;
+  ros::Duration elapsed_time = current_time - start_time;
+
+  for (int i = 0; i < data->name.size(); i++) {
+    std::string joint_name = data->name[i];
+
+    auto joint_it = std::find_if(joints.begin(), joints.end(), [&joint_name](auto j) { return j->getName() == joint_name; } );
+
+    if (joint_it != joints.end()) {
+      auto joint = *joint_it;
+
+      // TODO: SENDER ID!!!!! Generate this properly.
+      uint64_t sender_id = 1;
+
+      // Process motion control commands (if any)
+      auto p_cmd = std::numeric_limits<double>::quiet_NaN();
+      auto v_cmd = std::numeric_limits<double>::quiet_NaN();
+      auto e_cmd = std::numeric_limits<double>::quiet_NaN();
+      bool has_command = false;
+      if (i < data->position.size()) {
+        p_cmd = data->position[i];
+        has_command = true;
+      }
+      if (i < data->velocity.size()) {
+        v_cmd = data->velocity[i];
+        has_command = true;
+      }
+      if (i < data->effort.size()) {
+        e_cmd = data->effort[i];
+        has_command = true;
+      }
+      // TODO: verify nans are handled correctly; what about empty commands?
+      if (has_command) {
+        joint->setCommand(p_cmd, v_cmd, e_cmd, sender_id, command_lifetime/1000.0, current_time.toSec());
+      }
+
+      // Set name
+      if (i < data->settings.name.size()) {
+        joint->setName(data->settings.name[i]);
+      }
+
+      // Change control strategy
+      if (i < data->settings.control_strategy.size()) {
+        joint->setControlStrategy(static_cast<hebi::sim::Joint::ControlStrategy>(data->settings.control_strategy[i]));
+      }
+
+      // TODO: consider changing this to either direct setting of individual parameters,
+      // or an entire separate "optional" message layer...
+
+      // Change gains:
+      auto current_pos_gains = joint->getPositionPid().getGains();
+      if (updateGains(current_pos_gains, data->settings.position_gains, i))
+        joint->getPositionPid().setGains(current_pos_gains);
+
+      auto current_vel_gains = joint->getVelocityPid().getGains();
+      if (updateGains(current_vel_gains, data->settings.velocity_gains, i))
+        joint->getVelocityPid().setGains(current_vel_gains);
+
+      auto current_eff_gains = joint->getEffortPid().getGains();
+      if (updateGains(current_eff_gains, data->settings.effort_gains, i))
+        joint->getEffortPid().setGains(current_eff_gains);
+
+    }
+  }
+}
+
+//Service callback which acknowledges that a command has been received
+bool HebirosGazeboGroup::SrvAcknowledge(std_srvs::Empty::Request &req,
+  std_srvs::Empty::Response &res) {
+
+  this->check_acknowledgement = true;
+
+  if (this->acknowledgement) {
+    this->check_acknowledgement = false;
+    this->acknowledgement = false;
+    return true;
+  }
+  else {
```

Comment:
I'm pretty sure IntelliJ drilled this into my head (one of the few Java things which could carry over to here), but I try to remove the return from the `else{}` branch in cases like this.

More generically, it's in the context of:

```c++
SomeType someFunction(args...) {
  // stuff
  // ...
  if (someCondition) {
    // if true logic
    return someValueOnTrue;
  } else {
    // if false logic
  }
  // Note we are at the end now! IntelliJ or somebody else always complained when I didn't put the return here
  return someValueOnFalse;
}
```

(I don't really care too much about it though)

Reply:
Same here -- we agree on this style.  Yep -- I just didn't clear up this file.  Sorry; I should have let you know that pre-review.